File: src/query/router.ts
import type { QueryResult } from "../types.js";
import type { Connector } from "../connectors/base.js";
import { MySQLConnector } from "../connectors/mysql.js";
import { PostgresConnector } from "../connectors/postgres.js";
import { QueryEngine } from "./engine.js";
import { checkSqlSafety } from "./safety.js";

/**
 * Query router — decides whether to execute via MySQL/PostgreSQL directly or DuckDB.
 *
 * Routing logic:
 * - Pure DuckDB tables (Excel/CSV) → DuckDB engine
 * - Pure MySQL query (all referenced tables from same MySQL source) → MySQL direct
 * - Pure PostgreSQL query (all referenced tables from same PG source) → PostgreSQL direct
 * - Mixed / unknown → DuckDB (data already loaded at connect time)
 */
export class QueryRouter {
  private connectors: Map<string, Connector> = new Map();
  private engine: QueryEngine;
  // Map of table_name → source_id for quick routing lookup
  private tableSourceMap: Map<string, string> = new Map();

  constructor(engine: QueryEngine) {
    this.engine = engine;
  }

  registerConnector(sourceId: string, connector: Connector, tables: string[]): void {
    this.connectors.set(sourceId, connector);
    for (const t of tables) {
      this.tableSourceMap.set(t.toLowerCase(), sourceId);
    }
  }

  /**
   * Route and execute a query.
   */
  async route(sql: string, maxRows?: number): Promise<QueryResult> {
    const check = checkSqlSafety(sql);
    if (!check.safe) {
      throw new Error(`SQL safety check failed: ${check.reason}`);
    }

    const directSource = this.detectDirectSource(sql);
    if (directSource) {
      return this.executeOnDirectSource(directSource, sql, maxRows);
    }

    return this.engine.execute(sql, maxRows);
  }

  getEngine(): QueryEngine {
    return this.engine;
  }

  /**
   * Detect if all referenced tables in the SQL belong to the same direct-query source
   * (MySQL or PostgreSQL). Returns the source_id if yes, null otherwise.
   */
  private detectDirectSource(sql: string): string | null {
    // Extract table names from SQL (simple regex, covers common patterns)
    const tablePattern = /(?:FROM|JOIN)\s+[`"]?(\w+)[`"]?/gi;
    const mentioned: string[] = [];
    let match: RegExpExecArray | null;

    while ((match = tablePattern.exec(sql)) !== null) {
      mentioned.push(match[1].toLowerCase());
    }

    if (mentioned.length === 0) return null;

    // Check if all mentioned tables belong to the same direct-query source
    let candidateSource: string | null = null;
    for (const tableName of mentioned) {
      const sourceId = this.tableSourceMap.get(tableName);
      if (!sourceId) return null; // Unknown table, fall back to DuckDB
      if (!sourceId.startsWith("mysql:") && !sourceId.startsWith("postgresql:")) return null;

      if (candidateSource === null) {
        candidateSource = sourceId;
      } else if (candidateSource !== sourceId) {
        return null; // Tables from different sources — need DuckDB
      }
    }

    return candidateSource;
  }

  /**
   * Execute a query directly on the source database (MySQL or PostgreSQL)
   * and wrap result in QueryResult format.
   */
  private async executeOnDirectSource(
    sourceId: string,
    sql: string,
    maxRows?: number,
  ): Promise<QueryResult> {
    const connector = this.connectors.get(sourceId);
    if (!connector || (!(connector instanceof MySQLConnector) && !(connector instanceof PostgresConnector))) {
      // Fallback to DuckDB
      return this.engine.execute(sql, maxRows);
    }

    const limit = maxRows ?? 1000;
    // Append LIMIT if not already present
    const limitedSql = /\bLIMIT\s+\d+\s*$/i.test(sql.trim())
      ? sql
      : `${sql.trim()} LIMIT ${limit}`;

    const { columns, rows } = await connector.executeQuery(limitedSql);

    return {
      columns,
      rows,
      rowCount: rows.length,
      truncated: rows.length === limit,
    };
  }
}

File: lucid_skill/query/router.py
from __future__ import annotations

import re

from lucid_skill.connectors.base import Connector
from lucid_skill.query.engine import QueryEngine
from lucid_skill.query.safety import check_sql_safety
from lucid_skill.types import QueryResult


class QueryRouter:
    """Query router -- decides whether to execute via MySQL/PostgreSQL directly or DuckDB.

    Routing logic:
    - Pure DuckDB tables (Excel/CSV) -> DuckDB engine
    - Pure MySQL query (all referenced tables from same MySQL source) -> MySQL direct
    - Pure PostgreSQL query (all referenced tables from same PG source) -> PostgreSQL direct
    - Mixed / unknown -> DuckDB (data already loaded at connect time)
    """

    def __init__(self, engine: QueryEngine) -> None:
        self._connectors: dict[str, Connector] = {}
        self._engine = engine
        self._table_source_map: dict[str, str] = {}

    def register_connector(
        self, source_id: str, connector: Connector, tables: list[str]
    ) -> None:
        self._connectors[source_id] = connector
        for t in tables:
            self._table_source_map[t.lower()] = source_id

    def route(self, sql: str, max_rows: int | None = None) -> QueryResult:
        """Route and execute a query."""
        safe, reason = check_sql_safety(sql)
        if not safe:
            raise ValueError(f"SQL safety check failed: {reason}")

        direct_source = self._detect_direct_source(sql)
        if direct_source:
            return self._execute_on_direct_source(direct_source, sql, max_rows)

        return self._engine.execute(sql, max_rows)

    def get_engine(self) -> QueryEngine:
        return self._engine

    def _detect_direct_source(self, sql: str) -> str | None:
        """Detect if all referenced tables belong to the same direct-query source
        (MySQL or PostgreSQL). Returns the source_id if yes, None otherwise."""
        table_pattern = re.compile(r'(?:FROM|JOIN)\s+[`"]?(\w+)[`"]?', re.IGNORECASE)
        mentioned = [m.group(1).lower() for m in table_pattern.finditer(sql)]

        if not mentioned:
            return None

        candidate_source: str | None = None
        for table_name in mentioned:
            source_id = self._table_source_map.get(table_name)
            if not source_id:
                return None  # Unknown table, fall back to DuckDB
            if not source_id.startswith("mysql:") and not source_id.startswith(
                "postgresql:"
            ):
                return None

            if candidate_source is None:
                candidate_source = source_id
            elif candidate_source != source_id:
                return None  # Tables from different sources -- need DuckDB

        return candidate_source

    def _execute_on_direct_source(
        self, source_id: str, sql: str, max_rows: int | None = None
    ) -> QueryResult:
        """Execute a query directly on the source database (MySQL or PostgreSQL)
        and wrap result in QueryResult format."""
        connector = self._connectors.get(source_id)
        if connector is None or not hasattr(connector, "execute_query"):
            # Fallback to DuckDB
            return self._engine.execute(sql, max_rows)

        limit = max_rows if max_rows is not None else 1000
        # Append LIMIT if not already present
        if re.search(r'\bLIMIT\s+\d+\s*$', sql.strip(), re.IGNORECASE):
            limited_sql = sql
        else:
            limited_sql = f"{sql.strip()} LIMIT {limit}"

        result = connector.execute_query(limited_sql)  # type: ignore[attr-defined]
        columns = result["columns"]
        rows = result["rows"]

        return QueryResult(
            columns=columns,
            rows=rows,
            row_count=len(rows),
            truncated=len(rows) == limit,
        )
